Source/CesiumRuntime/Private/CesiumGltfComponent.cpp

```diff
@@ -2877,6 +2877,39 @@ static void SetMetadataParameterValues_DEPRECATED(
 PRAGMA_ENABLE_DEPRECATION_WARNINGS
 #pragma endregion
 
+namespace {
+void addInstanceFeatureIds(UCesiumGltfInstancedComponent* pInstancedComponent) {
```

Comment:
It adds storage for the feature IDs if there are any, otherwise I don't think there's any additional memory usage. Isn't this the same situation with regular feature IDs stored in textures?

Reply:
No, I think this is a lot different from the feature ID texture case. With attribute and texture feature IDs, space to hold a set of feature IDs is only allocated if that feature ID set is _used_ by a the CesiumFeaturesMetadata component. Step through `encodePrimitiveFeaturesAnyThreadPart` in `CesiumEncodedFeaturesMetadata.cpp` in the Metadata level in the Samples, and you'll see that `encodeFeatureIdTexture` is called. However, if you delete the CesiumFeaturesMetadata component from the "San Francisco Ferry Building" tileset, or if you remove the feature IDs from it, and then refresh the tileset, you'll see that that function is no longer called. The texture is only created if it's needed.

For instances, though, it appears that the "custom data" for the instance IDs is added, even if the tileset doesn't even have a CesiumFeaturesMetadata component. This means that all tilesets with instance IDs will use more memory than necessary (either on the CPU or GPU or both, I'm not sure where custom data values are stored) when those instance IDs aren't used at all.

Reply:
> No, I think this is a lot different from the feature ID texture case. With attribute and texture feature IDs, space to hold a set of feature IDs is only allocated if that feature ID set is _used_ by a the CesiumFeaturesMetadata component. Step through `encodePrimitiveFeaturesAnyThreadPart` in `CesiumEncodedFeaturesMetadata.cpp` in the Metadata level in the Samples, and you'll see that `encodeFeatureIdTexture` is called. However, if you delete the CesiumFeaturesMetadata component from the "San Francisco Ferry Building" tileset, or if you remove the feature IDs from it, and then refresh the tileset, you'll see that that function is no longer called. The texture is only created if it's needed.
> 
> For instances, though, it appears that the "custom data" for the instance IDs is added, even if the tileset doesn't even have a CesiumFeaturesMetadata component. This means that all tilesets with instance IDs will use more memory than necessary (either on the CPU or GPU or both, I'm not sure where custom data values are stored) when those instance IDs aren't used at all.

I understand now what you're saying; I wasn't thinking of the possibility of not loading model data that's part of the glTF. If I understand correctly, there's a big difference between the feature ID approach with normal models vs GPU instances: in the former, the feature IDs are stored in textures that are part of a material created during the encoding procedure, whereas instance feature IDs are stored in  `PerInstanceSMCustomData`, which is like a vertex attribute with a per-instance value. If the instance feature IDs are not stored when the static mesh is first created, then they would need to be modified during encoding...

The extra memory used is essentially one float per instance -- more accurately one float per feature ID set, though it seems rare to have more than one -- or about 10%. On the GPU.  Is this a blocker for you?

Reply:
I went ahead and added the logic to avoid populating the custom data for instance feature ID sets that aren't used in https://github.com/CesiumGS/cesium-unreal/pull/1565/commits/d63f692c49e065cfa518adbc038aa64cb6ce83ae. I appreciate your review on it when you get a chance, but I'm going to go ahead and merge this so that we can ship it on January 2!